src/processor/operator/recursive_extend/bfs_state.cpp

```diff
@@ -63,5 +44,32 @@ void ShortestPathBFSMorsel::resetVisitedState() {
     }
 }
 
+void ShortestPathTrackPathMorsel::markVisited(
+    common::offset_t boundOffset, common::offset_t nbrOffset, uint64_t multiplicity) {
+    assert(multiplicity == 1);
+    auto nextFrontierWithEdge = reinterpret_cast<FrontierWithEdge*>(nextFrontier);
+    if (visitedNodes[nbrOffset] == NOT_VISITED_DST) {
+        visitedNodes[nbrOffset] = VISITED_DST;
+        numVisitedDstNodes++;
+        nextFrontierWithEdge->addEdge(boundOffset, nbrOffset);
+    } else if (visitedNodes[nbrOffset] == NOT_VISITED) {
+        visitedNodes[nbrOffset] = VISITED;
+        nextFrontierWithEdge->addEdge(boundOffset, nbrOffset);
+    }
+}
+
+void ShortestPathNoTrackPathMorsel::markVisited(
```

Comment:
Looking at this function and the above also makes me thinking templating could be a better solution here.